File: Back Tracking/组合/78.子集.cpp
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <algorithm>
#include <unordered_set>
#include <unordered_map>
#include <math.h>

using namespace std;
/*
 * @lc app=leetcode.cn id=78 lang=cpp
 *
 * [78] 子集
 */

// @lc code=start
class Solution
{
public:
    vector<vector<int>> result; // 存放结果
    vector<int> path;
    void backtracking(vector<int> &nums, int startIndex)
    {
        result.push_back(path);
        // 可不写，因为startIndex超过数组长度时候，循环本身也结束了
        if (startIndex >= nums.size())
        {
            return;
        }

        for (int i = startIndex; i < nums.size(); i++)
        {
            path.push_back(nums[i]);
            // 这里是i+1
            backtracking(nums, i + 1);
            path.pop_back();
        }
    }
    vector<vector<int>> subsets(vector<int> &nums)
    {
        result.clear();
        path.clear();
        backtracking(nums, 0);
        return result;
    }
};
// @lc code=end

File: Back Tracking/组合/78.子集.ts
/*
 * @lc app=leetcode.cn id=78 lang=typescript
 *
 * [78] 子集
 */

// @lc code=start
function subsets(nums: number[]): number[][] {
  const result = []
  if (nums.length === 0) {
    return result
  }

  if (nums.length === 1) {
    return [[], nums]
  }

  // 数组升序排列
  nums = nums.sort((a, b) => a - b)
  const backTracking = (arr: number[], start: number) => {
    result.push([...arr])
    if (startIndex >= nums.size()) {
      return;
    }
    for (let i = start; i < nums.length; i++) {
      arr.push(nums[i])
      backTracking(arr, i + 1)
      arr.pop()
    }
  }

  backTracking([], 0)

  return result
};
// @lc code=end

